File: knowledge/langgraph/knowledge/store.py
"""Abstract knowledge store protocol."""

from __future__ import annotations

from typing import Protocol

from langgraph.knowledge.types import KnowledgeEntry, StoreStatus


class KnowledgeStore(Protocol):
    """Protocol that any knowledge-store backend must satisfy."""

    async def store(self, entry: KnowledgeEntry) -> KnowledgeEntry: ...

    async def search(
        self, query: str, *, limit: int = 10
    ) -> list[KnowledgeEntry]: ...

    async def get_by_topic(self, topic: str) -> list[KnowledgeEntry]: ...

    async def get_by_tags(self, tags: list[str]) -> list[KnowledgeEntry]: ...

    async def status(self) -> StoreStatus: ...

    async def delete(self, id: str) -> None: ...

File: knowledge-server/internal/store/store.go
// Package store defines the knowledge store interface and implementations.
package store

import "github.com/clab/knowledge-server/internal/types"

// KnowledgeStore is the interface that any knowledge-store backend must satisfy.
type KnowledgeStore interface {
	// Store persists a knowledge entry. Overwrites if the ID already exists.
	Store(entry types.KnowledgeEntry) (types.KnowledgeEntry, error)

	// Search performs keyword search and returns entries ranked by relevance.
	Search(query string, limit int) ([]types.KnowledgeEntry, error)

	// GetByTopic returns entries matching the topic exactly (case-insensitive).
	GetByTopic(topic string) ([]types.KnowledgeEntry, error)

	// GetByTags returns entries matching any of the given tags (case-insensitive).
	GetByTags(tags []string) ([]types.KnowledgeEntry, error)

	// Status returns a summary of the store state.
	Status() (types.StoreStatus, error)

	// Delete removes an entry by ID. Returns an error if not found.
	Delete(id string) error
}
